### PR TITLE
feat: add msm without serialization

### PR DESCRIPTION
**Pull request by diS3e**
Improved implementation of FFI `msm`, working without serialization to `ByteString`

### CODE REVIEW

**Comment by vlasin** [comment]
1. I propose bumping the size in the `BenchProve.hs` to `128` (the Haskell version will actually use FFT in that case).
2. Update README with the info about the new benchmarks.

File: rust-wrapper/Cargo.toml [hunk 1]
```diff
@@ -9,19 +9,20 @@ ark-ec = { version= "0.4.0" }
 ark-std = { version = "0.4.0", default-features = false }
 ark-serialize = { version = "0.4.0", default-features = false }
 ark-bls12-381 = { version = "0.4.0" }
-ark-test-curves = { version = "0.4.0", features= ["bls12_381_curve", "bls12_381_scalar_field"] }
-borsh          = "0.10"
+ark-msm = { git = "https://github.com/TalDerei/arkmsm.git", rev="bc95ea3784983d8ced03a642d765bbfdd91faa9b"}
```
Comment (TurtlePU) [inline rust-wrapper/Cargo.toml]: Why is it pinned by commit and not by a semver?

File: rust-wrapper/src/fft.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
+use ark_ff::PrimeField;
+use ark_poly::univariate::DensePolynomial;
+use ark_poly::DenseUVPolynomial;
+use libc;
+use num_bigint::BigUint;
+use std::ops::Mul;
+
+use crate::utils::deserialize_vector_scalar_field;
+
+fn pad_end(mut vec: Vec<u8>, target_length: usize) -> Vec<u8> {
```
Comment (TurtlePU) [inline rust-wrapper/src/fft.rs]: [`std::vec::Vec::resize`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.resize) can be used instead

File: rust-wrapper/src/fft.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        .to_vec()
+        .iter()
+        .map(|x| pad_end(BigUint::from(x.into_bigint()).to_bytes_le(), 32))
+        .collect();
+
+    unsafe {
+        std::ptr::copy(
+            res.join(&[][..]).as_ptr() as *const u8,
+            out as *mut u8,
+            res.len() * 32,
```
Comment (TurtlePU) [inline rust-wrapper/src/fft.rs]: Using [`std::mem::size_of`](https://doc.rust-lang.org/stable/std/mem/fn.size_of.html) can be better both for self-documentation and to reduce bugs

File: rust-wrapper/src/fft.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    out: *mut libc::c_char,
+) {
+    let l = DensePolynomial::from_coefficients_vec(deserialize_vector_scalar_field(l_var, l_len));
+    let r = DensePolynomial::from_coefficients_vec(deserialize_vector_scalar_field(r_var, r_len));
+
+    let res: Vec<Vec<u8>> = l
+        .mul(&r)
+        .to_vec()
+        .iter()
+        .map(|x| pad_end(BigUint::from(x.into_bigint()).to_bytes_le(), 32))
```
Comment (TurtlePU) [inline rust-wrapper/src/fft.rs]: Here `32` is also actually [`std::mem::size_of`](https://doc.rust-lang.org/stable/std/mem/fn.size_of.html), I guess

File: rust-wrapper/src/msm.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        .collect();
+
+    let points = deserialize_vector_points(points_var, points_len);
+
+    let opt_window_size = get_opt_window_size(log2(points.len()));
+    let r: GAffine = VariableBaseMSM::multi_scalar_mul_custom(
+        &points,
+        &scalars,
+        opt_window_size,
+        2048,
```
Comment (TurtlePU) [inline rust-wrapper/src/msm.rs:65]: What is the meaning of `2048` and `256` here?

File: rust-wrapper/src/msm.rs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        marshall_from_haskell_var::<RW, Buffer<ScalarField>>(scalars_var, scalars_len, RW).0;
+    let points: Vec<GAffine> =
+        marshall_from_haskell_var::<RW, Buffer<GAffine>>(points_var, points_len, RW).0;
+
+    let r: GAffine = G::msm(&points, &scalars).unwrap().into();
+
+    marshall_to_haskell_var(&r, out, out_len, RW);
+}
+
+const fn get_opt_window_size(k: u32) -> u32 {
```
Comment (TurtlePU) [inline rust-wrapper/src/msm.rs:38]: How is this function derived? What is the meaning of these intervals and constants?

File: rust-wrapper/src/msm.rs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+#[no_mangle]
+pub extern "C" fn rust_wrapper_multi_scalar_multiplication(
+    points_var: *const u8,
+    points_len: usize,
+    scalars_var: *const u8,
+    scalars_len: usize,
+    out: *mut u8,
+    out_len: &mut usize,
+) {
+    let scalars: Vec<ScalarField> =
```
Comment (TurtlePU) [inline rust-wrapper/src/msm.rs]: `rustc` should be able to infer these type annotations, can it not?

File: rust-wrapper/src/fft.rs (hunk 4 above)
Comment (TurtlePU) [reply rust-wrapper/src/fft.rs]: also [`std::iter::Iterator::flatMap`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.flat_map) can be used to flatten nested structure

File: rust-wrapper/src/msm.rs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    points_var: *const libc::c_char,
+    points_len: usize,
+    scalars_var: *const libc::c_char,
+    scalars_len: usize,
+    out_len: usize,
+    out: *mut libc::c_char,
+) {
+    let scalars: Vec<_> = deserialize_vector_scalar_field(scalars_var, scalars_len)
+        .iter()
+        .map(|i| {
```
Comment (TurtlePU) [inline rust-wrapper/src/msm.rs]: `{ return expr; }` can be shortened to just `expr`, so here it's just `|i| i.into_bigint()` (or, if possible, `ScalarField::into_bigint`)

File: rust-wrapper/src/utils.rs [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+pub fn deserialize_vector<T>(
+    vector: &mut Vec<u8>,
+    object_size: usize,
+    deserialize: fn(&mut [u8]) -> T,
+) -> Buffer<T> {
+    Buffer(
+        vector
+            .chunks_exact_mut(object_size)
+            .map(|chunk| deserialize(chunk))
```
Comment (TurtlePU) [inline rust-wrapper/src/utils.rs]: This can be just `deserialize`

File: rust-wrapper/src/utils.rs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        std::ptr::copy(scalars_var as *const u8, vec.as_mut_ptr(), scalars_len);
+    }
+    let scalars: Vec<_> = deserialize_vector(&mut vec, 32, |x| -> ScalarField {
+        PrimeField::from_le_bytes_mod_order(x)
+    })
+    .0;
+    scalars
+}
+
+pub fn deserialize_vector_points(
```
Comment (TurtlePU) [inline rust-wrapper/src/utils.rs]: Are we sure this function is safe?

File: rust-wrapper/src/utils.rs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+) -> Buffer<T> {
+    Buffer(
+        vector
+            .chunks_exact_mut(object_size)
+            .map(|chunk| deserialize(chunk))
+            .collect(),
+    )
+}
+
+pub fn deserialize_vector_scalar_field(
```
Comment (TurtlePU) [inline rust-wrapper/src/utils.rs]: Are we sure this function is safe?

File: rust-wrapper/src/utils.rs [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+            PrimeField::from_le_bytes_mod_order(bytes)
+        });
+        Ok(res)
+    }
+}
+
+impl FromHaskell<RW> for Buffer<GAffine> {
+    fn from_haskell(buf: &mut &[u8], tag: PhantomData<RW>) -> Result<Self> {
+        let mut buffer = <Vec<u8>>::from_haskell(buf, tag)?;
+        let res = deserialize_vector(&mut buffer, 96, |bytes| -> GAffine {
```
Comment (TurtlePU) [inline rust-wrapper/src/utils.rs]: `96` looks like [`std::mem::size_of`](https://doc.rust-lang.org/stable/std/mem/fn.size_of.html)

File: rust-wrapper/src/utils.rs [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        let mut res = Vec::new();
+        self.serialize_uncompressed(&mut res).unwrap();
+        res.to_haskell(writer, tag)
+    }
+}
+
+impl FromHaskell<RW> for Buffer<ScalarField> {
+    fn from_haskell(buf: &mut &[u8], tag: PhantomData<RW>) -> Result<Self> {
+        let mut buffer = <Vec<u8>>::from_haskell(buf, tag)?;
+        let res = deserialize_vector(&mut buffer, 32, |bytes| -> ScalarField {
```
Comment (TurtlePU) [inline rust-wrapper/src/utils.rs]: `32` looks like [`std::mem::size_of`](https://doc.rust-lang.org/stable/std/mem/fn.size_of.html)

File: rust-wrapper/src/utils.rs [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        let _a = <Vec<u8>>::from_haskell(buf, tag)?;
+        let a = GAffine::deserialize_uncompressed(&*_a).unwrap();
+        Ok(a)
+    }
+}
+
+impl ToHaskell<RW> for GAffine {
+    fn to_haskell<W: Write>(&self, writer: &mut W, tag: PhantomData<RW>) -> Result<()> {
+        let mut res = Vec::new();
+        self.serialize_uncompressed(&mut res).unwrap();
```
Comment (TurtlePU) [inline rust-wrapper/src/utils.rs]: It is better to propagate the error via `?` instead of calling `.unwrap()`

File: rust-wrapper/src/utils.rs [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        BigUint::from(self.into_bigint())
+            .to_bytes_le()
+            .to_haskell(writer, tag)
+    }
+}
+
+impl FromHaskell<RW> for GAffine {
+    fn from_haskell(buf: &mut &[u8], tag: PhantomData<RW>) -> Result<Self> {
+        let _a = <Vec<u8>>::from_haskell(buf, tag)?;
+        let a = GAffine::deserialize_uncompressed(&*_a).unwrap();
```
Comment (TurtlePU) [inline rust-wrapper/src/utils.rs]: It is better to propagate the error via `?` instead of calling `.unwrap()`

File: rust-wrapper/src/fft.rs [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    vec
+}
+
+#[no_mangle]
+pub extern "C" fn rust_wrapper_mul_fft(
+    l_var: *const libc::c_char,
+    l_len: usize,
+    r_var: *const libc::c_char,
+    r_len: usize,
+    _out_len: usize,
```
Comment (TurtlePU) [inline rust-wrapper/src/fft.rs]: Why is `_out_len` not used?

File: rust-wrapper/src/fft.rs (hunk 3 above)
Comment (TurtlePU) [reply rust-wrapper/src/fft.rs]: (also maybe using `_out_len` here is better)

**Comment by TurtlePU** [comment]
:fire: :fire: :fire:

File: rust-wrapper/src/fft.rs (hunk 16 above)
Comment (diS3e) [reply rust-wrapper/src/fft.rs]: fixed

File: rust-wrapper/src/utils.rs (hunk 15 above)
Comment (diS3e) [reply rust-wrapper/src/utils.rs]: fixed

File: rust-wrapper/src/utils.rs (hunk 14 above)
Comment (diS3e) [reply rust-wrapper/src/utils.rs]: fixed

File: rust-wrapper/src/utils.rs (hunk 13 above)
Comment (diS3e) [reply rust-wrapper/src/utils.rs]: fixed

File: rust-wrapper/src/utils.rs (hunk 12 above)
Comment (diS3e) [reply rust-wrapper/src/utils.rs]: fixed

File: rust-wrapper/src/utils.rs (hunk 11 above)
Comment (diS3e) [reply rust-wrapper/src/utils.rs]: fixed

File: rust-wrapper/src/utils.rs (hunk 10 above)
Comment (diS3e) [reply rust-wrapper/src/utils.rs]: fixed

File: rust-wrapper/src/utils.rs (hunk 9 above)
Comment (diS3e) [reply rust-wrapper/src/utils.rs]: fixed

File: rust-wrapper/src/msm.rs (hunk 8 above)
Comment (diS3e) [reply rust-wrapper/src/msm.rs]: fixed

File: rust-wrapper/src/msm.rs (hunk 7 above)
Comment (diS3e) [reply rust-wrapper/src/msm.rs]: fixed

File: rust-wrapper/src/msm.rs (hunk 6 above)
Comment (diS3e) [reply rust-wrapper/src/msm.rs:38]: this is inner function from `arkmsm` crate. 
It is not public, but is used in the `multi_scalar_mul_custom` call. It was probably expected that the wrapper method `multi_scalar_mul` would be called, but `multi_scalar_mul_g1_glv` (one of the options in `multi_scalar_mul`) does not work correctly at the moment

File: rust-wrapper/src/msm.rs (hunk 5 above)
Comment (diS3e) [reply rust-wrapper/src/msm.rs:65]: this also from `arkmsm` crate:
https://github.com/snarkify/arkmsm/blob/main/src/msm.rs

File: rust-wrapper/src/fft.rs (hunk 4 above)
Comment (diS3e) [reply rust-wrapper/src/fft.rs]: fixed

File: rust-wrapper/src/fft.rs (hunk 3 above)
Comment (diS3e) [reply rust-wrapper/src/fft.rs]: fixed

File: rust-wrapper/src/fft.rs (hunk 2 above)
Comment (diS3e) [reply rust-wrapper/src/fft.rs]: fixed

File: rust-wrapper/src/msm.rs (hunk 5 above)
Comment (TurtlePU) [reply rust-wrapper/src/msm.rs:65]: Well, these look like sizes of some types, so maybe `size_of` is better to use here

File: rust-wrapper/src/msm.rs (hunk 6 above)
Comment (TurtlePU) [reply rust-wrapper/src/msm.rs:38]: Ok! Then a comment with a link to the reference implementation would be a nice addition

File: rust-wrapper/src/msm.rs (hunk 5 above)
Comment (diS3e) [reply rust-wrapper/src/msm.rs:65]: I don't see any connection between these constants and sizes of the types. Apparently these is just optimal constants and they are not associated with any objects

File: .github/workflows/main-pull.yml
Comment (TurtlePU) [inline .github/workflows/main-pull.yml:1]: It's better to add CI in a separate PR

File: haskell-wrapper/tests/MultiScalarMultiplication.hs [hunk 17]
```diff
@@ -17,7 +17,7 @@ specMultiScalarMultiplication :: IO ()
 specMultiScalarMultiplication = hspec $ do
     describe "Rust binary scalar mulitply specification" $ do
         it "should be equal to haskell" $ do
-            property $ withMaxSuccess 200 $
+            property $ withMaxSuccess 100 $
```
Comment (TurtlePU) [inline haskell-wrapper/tests/MultiScalarMultiplication.hs]: 100 is a default, you don't have to specify it

File: rust-wrapper/src/fft.rs [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+use libc;
+use num_bigint::BigUint;
+use std::ops::Mul;
+
+use crate::utils::deserialize_vector_scalar_field;
+
+///
+/// # Safety
+/// The caller must ensure that valid pointers and sizes are passed. 
+/// To find out how data should be represented in memory, you can look at fft_bench.rs
```
Comment (TurtlePU) [inline rust-wrapper/src/fft.rs]: You can extract the safe part of a function which turns slice of bytes into polynomials and returns vector of bytes, and use the safe part in benchmarks, while leaving the unsafe parts to FFI

File: rust-wrapper/src/msm.rs [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    }
+}
+
+///
+/// # Safety
+/// The caller must ensure that valid pointers and sizes are passed. 
+/// To find out how data should be represented in memory, you can look at msm_bench.rs
+/// .
+#[no_mangle]
+pub unsafe extern "C" fn rust_wrapper_multi_scalar_multiplication_without_serialization(
```
Comment (TurtlePU) [inline rust-wrapper/src/msm.rs]: The most part of the function is safe, once again it is better to extract the safe part into separate safe function and leave the unsafe part in FFI

File: rust-wrapper/src/fft.rs (hunk 18 above)
Comment (TurtlePU) [reply rust-wrapper/src/fft.rs]: Also I guess there should be a no-op (unsafe) conversion from the (pointer, size) pair into a Rust slice, and it's better to use them

File: rust-wrapper/src/msm.rs [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+use ark_msm::msm::VariableBaseMSM;
+use ark_serialize::CanonicalSerialize;
+use ark_std::log2;
+use haskell_ffi::{from_haskell::marshall_from_haskell_var, to_haskell::marshall_to_haskell_var};
+use utils::{Buffer, RW};
+
+use crate::utils::{self, deserialize_vector_points, deserialize_vector_scalar_field};
+
+#[no_mangle]
+pub extern "C" fn rust_wrapper_multi_scalar_multiplication(
```
Comment (TurtlePU) [inline rust-wrapper/src/msm.rs]: This function also looks unsafe

**Review by TurtlePU — APPROVED** [review]

**Review by vlasin — APPROVED** [review]

**Comment by vlasin** [comment]
Great job! :fire:

**Pull request merged**